### PR TITLE
feat(benchmark): HSTU E2E training benchmark suite with progressive optimizations

### LINKED ISSUE #307: [FEA] Overlap DP and MP embedding.
Currently, due to **D2H and stream sync**, the dynamicemb( `MP`) and dense (torchrec DP) is serialized, the [side stream](https://github.com/NVIDIA/recsys-examples/blob/main/examples/commons/modules/embedding.py#L397-L404) takes no any effect.

See below timeline

Image

### PR DESCRIPTION
**Pull request by JacoCheung**
## Summary

Add a comprehensive HSTU end-to-end training benchmark suite that measures the impact of progressive optimizations on H100 GPUs.

- **Benchmark infrastructure**: automated experiment generation (`generate_gin_config.py`), SLURM submission scripts, result analysis and visualization tools
- **5 progressive experiments**: Baseline → Workload-Balanced Shuffler → CUTLASS Attention → Selective Recompute → Tensor Parallel (TP=2)
- **Results on 2×H100 nodes (16 GPUs)**: Baseline 1092 TFLOPS (6.38% MFU) → Best 3933 TFLOPS (22.96% MFU), **3.60× speedup**
- **CUTLASS attention kernel MFU heatmap**: batch sizes 1–128 × seqlens 64–4096, peak ~60% MFU on H100

### Key code changes (non-benchmark)

- **perf**: eliminate D2H sync in DP embedding forward and enable DP/MP overlap (#307)
- **perf**: avoid D2H sync in `_Split2DJaggedFunction` by precomputing split lengths
- **fix**: triton attention correctness for `num_contextuals`, empty batch handling
- **fix(tp)**: scale `total_candidates_seq_len` for TP and shuffler compatibility
- **fix**: shuffler race condition, batch counter, batch all2all support
- **fix**: `num_loss_tokens()` added to `BaseBatch`/`GPTSIDBatch`, sid_gr `progress()` signature updated
- **fix**: watchdog log decoupled from `MEM_DEBUG`
- **fix**: `batch_allgather` dense tensor padding guard for pre-padded tensors (#361)
- **style**: apply pre-commit formatting (isort 5.12.0, black, autoflake, codespell)

### Benchmark experiments

| # | Experiment | TFLOPS | MFU (%) | Speedup |
|---|-----------|--------|---------|---------|
| 0 | Baseline | 1092 | 6.38 | 1.00× |
| 1 | +Shuffler | 1667 | 9.73 | 1.53× |
| 2 | **+CUTLASS** | **3933** | **22.96** | **3.60×** |
| 3 | +Recompute | 3919 | 22.88 | 3.59× |
| 4 | +TP=2 | 2880 | 16.81 | 2.64× |

### Known Issues

- `pad_and_allgather_batch` dense tensor padding assumes unpadded dim-0, but some fields (e.g. `num_candidates`) are pre-padded by dataloader → #361

## Test plan

- [x] E2E benchmark on EOS (2×H100 nodes, zipf α=1.05) — 3933 TFLOPS peak, 22.96% MFU
- [x] CUTLASS kernel benchmark completed on H100
- [x] DP/MP overlap fix validated on H100 (movielen + benchmark configs)
- [x] `num_loss_tokens()` added for SID-GR compatibility
- [x] 4-GPU movielen ranking integration test (with workaround for #361)
- [ ] CI passes on PR

Closes #307

## CI

- [Pipeline #48364247](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/pipelines/48364247) — 2026-04-13 02:00 UTC
- [Pipeline #48369512](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/pipelines/48369512) — 2026-04-13 04:05 UTC
- [Pipeline #48373677](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/pipelines/48373677) — 2026-04-13 04:58 UTC
- [Pipeline #48381330](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/pipelines/48381330) — 2026-04-13 05:30 UTC (fix test_pipeline)
- [Pipeline #48381987](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/pipelines/48381987) — 2026-04-13 05:45 UTC (fix pre-commit formatting)
- [Pipeline #48383030](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/pipelines/48383030) — 2026-04-13 06:10 UTC (fix codespell + watchdog)
- [Pipeline #48385495](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/pipelines/48385495) — 2026-04-13 07:00 UTC (fix isort CI mismatch)
- [Pipeline #48391585](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/pipelines/48391585) — 2026-04-13 10:01 UTC (canceled)
- [Pipeline #48392513](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/pipelines/48392513) — 2026-04-13 10:15 UTC (canceled)
- [Pipeline #48408863](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/pipelines/48408863) — 2026-04-13 11:30 UTC (fix sid_gr training.py)
- [Pipeline #48410148](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/pipelines/48410148) — 2026-04-13 12:00 UTC (on CI fix branch)
- [Pipeline #48480041](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/pipelines/48480041) — 2026-04-14 (rebased on main)
- [Pipeline #48485296](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/pipelines/48485296) — 2026-04-14 (pre-commit fix for rebase conflict)

### CODE REVIEW

File: examples/hstu/training/benchmark/figs/hstu_attn_mfu.png
Comment (shijieliu) [inline examples/hstu/training/benchmark/figs/hstu_attn_mfu.png:1]: update benhcmark seqlen from 128 to 16k

File: examples/hstu/training/benchmark/E2E_BENCHMARK.md [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+
+<img src="figs/comparison.png" width="100%" />
+
+### Key Takeaways
+
+1. **CUTLASS attention is the foundation**: Replacing the Triton kernel with CUTLASS yields a 2.6× speedup — by far the most impactful single optimization, reflecting the attention-bound nature of HSTU.
+
+2. **Workload-balanced shuffler delivers the highest MFU (21.7%)**: Zipf-distributed sequence lengths cause severe load imbalance with O(n²) attention. Redistributing sequences to equalize per-GPU FLOPs eliminates idle time and adds another 1.36× on top of CUTLASS+Caching+Recompute.
+
+3. **Caching and recompute are memory-oriented**: DynamicEmb caching and selective recompute do not improve raw throughput (MFU drops slightly from 16.40% to 15.99%), but they significantly reduce memory pressure — enabling larger models or longer sequences.
```
Comment (shijieliu) [inline examples/hstu/training/benchmark/E2E_BENCHMARK.md]: this needs to be updated when benchmark results update

File: examples/commons/utils/logger.py [hunk 2]
```diff
@@ -13,6 +13,10 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 import logging
+
+# Set up logger with RichHandler if not already configured
+# Control via environment variable:
+#   DISABLE_RICH=1  - Disable Rich, use plain print instead
```
Comment (shijieliu) [inline examples/commons/utils/logger.py:19]: what's this?

Comment (JacoCheung) [reply examples/commons/utils/logger.py:19]: Rich is used to beautify the output to terminal.

The output of benchmark is redirected to file, so no need to use Rich.

File: examples/commons/distributed/batch_shuffler.py [hunk 3]
```diff
@@ -338,7 +333,7 @@ def finish_shuffle(
             )
```
[the remark this answers is not in the document]
Comment (JacoCheung) [reply examples/commons/distributed/batch_shuffler.py:333]: Not fixing: this only affects the log print window (which batch indices trigger logging). The handle carries the correct batch index for the log message itself. In a prefetch pipeline with multiple in-flight batches, the counter offset is expected and benign.

File: examples/commons/distributed/batch_shuffler.py [hunk 4]
```diff
@@ -137,8 +138,9 @@ def _log_load_balance(
```
[the remark this answers is not in the document]
Comment (JacoCheung) [reply examples/commons/distributed/batch_shuffler.py:139]: Not fixing: Python augmented assignment `self._batch_counter += 1` creates an instance attribute on first write, so each instance gets its own counter. This is standard Python behavior and works correctly.

File: examples/hstu/training/benchmark/figs/hstu_attn_mfu.png
Comment (JacoCheung) [reply examples/hstu/training/benchmark/figs/hstu_attn_mfu.png:1]: Done — updated in latest push.

File: examples/hstu/training/benchmark/E2E_BENCHMARK.md (hunk 1 above)
Comment (JacoCheung) [reply examples/hstu/training/benchmark/E2E_BENCHMARK.md]: Done — results table and key takeaways updated.

File: examples/commons/pipeline/train_pipeline.py [hunk 5]
```diff
@@ -758,11 +764,15 @@ def progress(self, dataloader_iter: Iterator[In]) -> Out:
...
 
+        global_tokens = None
         if self._model.training:
             with nvtx.annotate("## zero_grad ##"):
                 if hasattr(self._model.module, "zero_grad_buffer"):
                     self._model.module.zero_grad_buffer()
                 self._optimizer.zero_grad()
+            with nvtx.annotate("## global_tokens ##"):
+                global_tokens = self.batches[0].num_loss_tokens().to(self._device)
+                torch.distributed.all_reduce(global_tokens)
```
[the remark this answers is not in the document]
Comment (JacoCheung) [reply examples/commons/pipeline/train_pipeline.py:775]: This is expected considering our data input is similar to FSDP. (All ranks have unique input)

**Review by shijieliu — APPROVED** [review]
